### PR TITLE
hyperkube: add cloud-controller-manager

### LINKED ISSUE #55732: hyperkube should include cloud-controller-manager
**Is this a BUG REPORT or FEATURE REQUEST?**: Both

/kind feature

**What happened**:
I tried to do `hyperkube cloud-controller-manager`

**What you expected to happen**:
CCM to execute.

**How to reproduce it (as minimally and precisely as possible)**:
Try to run CCM from Hyperkube.

**Anything else we need to know?**:

I know there's a desire to move CCM, etc out of core, but I think this is an appropriate action to take for 2-3 releases, and I think it can encourage additional testing from folks that already have infra/installers that just operate with a single given hyperkube image/tag.

**Environment**: n/a

### PR DESCRIPTION
**Pull request by colemickens**
**What this PR does / why we need it**:

Adds `cloud-controller-manager` to `hyperkube`. (fix #55732)

This is useful as a number of deployment tools run all of the kubernetes components from the `hyperkube` image/binary. It also makes testing easier as a single binary/image can be built and pushed quickly.

This PR follows the same pattern of the other kubernetes binaries being available as part of `hyperkube`.

(This PR also makes an error condition appropriately fatal.)

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: n/a

**Special notes for your reviewer**: n/a

**Release note**:

```release-note
hyperkube: add cloud-controller-manager
```

/sig cluster-lifecycle
/area cloudprovider

### CODE REVIEW

**Review by luxas — COMMENTED** [review]
I might be too strict, but I´m not sure I wanna see the ccm being in hyperkube. I think that goes the wrong way (adding more deps on clouds rather than less)

The preferred way for clouds to do this should be to create their own binary with their own extensions ASAP. The ccm binary in-core is just a stop-gap.

That said, I see why it´s useful in the meantime, and we already build the binary in releases, creating images, etc. so we should probably be able to do this too.

**Comment by colemickens** [comment]
@luxas I agree with both of your points. I assume that when CCM is removed from the tree, we would also remove it from `hyperkube`. In the meantime, this can be helpful for a couple releases. Given how big the binary and image already are, hopefully this can be tolerable?

**Comment by luxas** [comment]
> @luxas I agree with both of your points. I assume that when CCM is removed from the tree, we would also remove it from hyperkube. In the meantime, this can be helpful for a couple releases. Given how big the binary and image already are, hopefully this can be tolerable?

Is there any way we could make this as `hyperkube alpha cloud-controller-manager` or something? That would alleviate the concerns on the hyperkube CLI API having to be stable for me at least

**Comment by colemickens** [comment]
Sure, let me take a look at doing so. Is there an existing example of this already? Thanks!

**Comment by colemickens** [comment]
@luxas I hacked in an `alpha` subcommand and added tests for it. Let me know what you think.

File: cmd/hyperkube/hyperkube.go [hunk 1]
```diff
@@ -154,7 +170,23 @@ func (hk *HyperKube) Run(args []string, stopCh <-chan struct{}) error {
...
 	}
 
-	s, err := hk.FindServer(serverName)
+	var s *Server
+	var err error
+	if serverName == "alpha" {
+		if len(args) > 0 && len(args[0]) > 0 {
+			serverName = args[0]
+			args = args[1:]
+			hk.Printf("Warning: alpha command syntax is unable!\n\n")
```
Comment (luxas) [inline cmd/hyperkube/hyperkube.go]: is unable to _what_?

File: cmd/hyperkube/cloud-controller-manager.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
+/*
+Copyright 2015 The Kubernetes Authors.
```
Comment (luxas) [inline cmd/hyperkube/cloud-controller-manager.go]: nit: 2017 :)

**Review by luxas — APPROVED** [review]
Kind of LGTM, left two nits

I'd like @wlan0 and @thockin sign off as well
/assign @wlan0 @thockin

File: cmd/hyperkube/hyperkube.go (hunk 1 above)
Comment (colemickens) [reply cmd/hyperkube/hyperkube.go]: Took me a second, it's supposed to be an instability warning. `s/unsable/unstable/g`. Will push a new commit and can squash at the end.

**Comment by luxas** [comment]
Thanks! Handing over to @wlan0 and @thockin then.

File: cmd/hyperkube/main.go [hunk 3]
```diff
@@ -46,5 +46,11 @@ func main() {
 	hk.AddServer(NewKubeProxy())
 	hk.AddServer(NewKubeAggregator())
 
+	// Alpha servers
+	hk.AddAlphaServer(NewCloudControllerManager())
+
+	// Alpha servers
+	hk.AddAlphaServer(NewCloudControllerManager())
```
Comment (wlan0) [inline cmd/hyperkube/main.go]: Should this be here twice?

**Comment by wlan0** [comment]
@colemickens "alpha" command wiring is cool. I just have one comment.

**Comment by thockin** [comment]
Assign back to me once it is LGTM'ed and ready for approval.

File: cmd/hyperkube/main.go (hunk 3 above)
Comment (colemickens) [reply cmd/hyperkube/main.go]: No, it shouldn't. Maybe a rebase error? Fixed in a new commit. (Will squash)

**Comment by colemickens** [comment]
@wlan0 Thanks for the review. Fixed the issue you caught and pushed a new commit. If it looks good, I'll squash.

**Comment by luxas** [comment]
ready for approval @thockin

**Comment by thockin** [comment]
This needs an associated issue and a `closes #12345` in the FIRST COMMENT, or it can not be approved.

**Comment by dims** [comment]
@thockin This is ready now with a link to a bug as you pointed out. PTAL

**Pull request merged**